### PR TITLE
NTP-based internet test

### PR DESCRIPTION
**Pull request by avtolstoy**
### Problem

1. Gen 3 platforms do not have internet test at all
2. Using 8.8.8.8:53 over TCP may not be ideal as it's TCP and most devices nowadays use UDP for cloud comms and it skips DNS step
3. The logic around resetting network interfaces due to a number of failed cloud connection attempts is broken and only works somewhat correctly with TCP-based devices (Photon and P1 only)

### Solution

This PR introduces an SNTP (Simple Network Time Protocol) implementation to be used for internet tests. By default we are targeting `pool.ntp.org`, which allows us to also test DNS and allows us to use the distributed network of NTP servers around the globe in the pool over UDP.

We are also changing the logic around when to trigger network interface reset due to cloud connection errors. Previously only "connection" attempts were registered as failures, whereas handshake errors were not taken into account at all. This does work somewhat ok for TCP-based devices (Photon/P1), but these days does not make much sense for connection-less UDP-based devices.

Now, we are taking into account any kind of a cloud connection error and will wait up to 5 minutes of consecutive cloud connection failures across all platforms to trigger a reset of network interfaces.

### Steps to Test

- Unit tests
- Corrupt device/server keys, watch the logs. The device should reset network interfaces after about 5 minutes
- Probably easiest to test on WiFi devices: disconnect WAN on router/set firewall rules to forbid the devices to communicate over UDP. The device should reset network interfaces after about 5 minutes

### Example App

N/A

### References

- [CH36520]
- [CH48170]
- [CH51505]
- [CH34989]

---

### Completeness

- [x] User is totes amazing for contributing!
- [ ] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [ ] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

**Comment by keeramis** [comment]
I will start by testing out some tests mentioned on one of my devices.

File: test/unit_tests/hal/simple_ntp_client/posix/simple_ntp_client.cpp [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+            SimpleNtpClient client;
+            uint64_t ntpTime = 0;
+            mocks.enableResponse(false);
+            CHECK(client.ntpDate(&ntpTime, "ipv4", 10000) == SYSTEM_ERROR_TIMEOUT);
+            CHECK(mocks.requestsReceived() == (10000 / ntp::RETRY_INTERVAL));
+        }
+        CHECK(mocks.socketClosed());
+        CHECK(mocks.addrInfoFreed());
+    }
+}
```
Comment (sergeuz) [inline test/unit_tests/hal/simple_ntp_client/posix/simple_ntp_client.cpp:297]: :+1:

File: hal/network/util/simple_ntp_client_detail.h [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    return val;
+#else
+    return __builtin_bswap32(val);
+#endif
+}
+
+inline uint32_t ntohl(uint32_t val) {
+    return htonl(val);
+}
+#endif // !HAL_USE_SOCKET_HAL_POSIX
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/util/simple_ntp_client_detail.h]: Fixed.

File: hal/network/util/simple_ntp_client.cpp [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+        if (r) {
+            if (s != sock_ || close_) {
+                sock_close(s);
+            }
+            continue;
+        }
+
+        sock = s;
+        LOG_DEBUG(TRACE, "SNTP connected");
+    }
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/util/simple_ntp_client.cpp:154]: Renamed to `isSockExternal_`

File: hal/network/util/simple_ntp_client.cpp [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+        sockaddr_t remote = {};
+        socklen_t remoteLen = sizeof(remote);
+        ssize_t r = socket_receivefrom(sock_, buf, size, 0, &remote, &remoteLen);
+        if (r >= 0) {
+            if (remote.sa_family == addr_.sa_family &&
+                    !memcmp(remote.sa_data, addr_.sa_data, sizeof(uint32_t) + sizeof(uint16_t))) {
+                return r;
+            }
+        }
+        HAL_Delay_Milliseconds(1);
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/util/simple_ntp_client.cpp]: Changed to 10ms, although I'd argue that it should be fine at 1ms as well :)

File: hal/network/util/simple_ntp_client.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace particle {
+
+// Forward declaration
+class UdpSocket;
+
+class SimpleNtpClient {
+public:
+    SimpleNtpClient();
+    SimpleNtpClient(sock_handle_t sock);
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/util/simple_ntp_client.h]: Fixed

File: hal/network/util/simple_ntp_client.cpp [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+    for (unsigned retry = 0; retry < retries; retry++) {
+        // Construct a request
+        ntp::Message msg = {};
+        msg.liVnMode = ntp::MODE_CLIENT | ntp::VERSION;
+        msg.originTimestamp = ntp::Timestamp::fromUnixtime(getRtcTime());
+
+        LOG_DEBUG(TRACE, "Sending request");
+        lastError = sock_->send((const uint8_t*)&msg, sizeof(msg));
+        if (lastError < 0) {
+            CHECK(sock_->connect(hostname, ntp::PORT));
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/util/simple_ntp_client.cpp]: Removed.

File: hal/network/util/simple_ntp_client.cpp [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+            LOG(WARN, "Failed to parse NTP response");
+        }
+    }
+
+    return lastError;
+}
+
+void ntp::dumpNtpTime(uint64_t timestamp) {
+    time_t unixTime = timestamp / USEC_IN_SEC;
+    unsigned usecs = timestamp - unixTime;
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/util/simple_ntp_client.cpp]: Added forgotten `* USEC_IN_SEC`

File: hal/network/util/simple_ntp_client_detail.h [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+
+struct Timestamp {
+    uint32_t seconds;
+    uint32_t fraction;
+
+    static Timestamp fromUnixtime(uint64_t unixMicros) {
+        uint32_t seconds = unixMicros / USEC_IN_SEC;
+        uint32_t frac = ((unixMicros - seconds) << FRACTIONS_IN_SEC_POW) / USEC_IN_SEC;
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/util/simple_ntp_client_detail.h]: Added forgotten `* USEC_IN_SEC`

File: hal/network/util/simple_ntp_client_detail.h [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        particle::Random rand;
+        rand.gen((char*)&frac, sizeof(char));
+        return {htonl(seconds + SECONDS_FROM_1970_TO_1900), htonl(frac)};
+    }
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/util/simple_ntp_client_detail.h:93]: Added comments.

File: system/src/system_task.cpp [hunk 10]
```diff
@@ -16,6 +16,10 @@
  ******************************************************************************
  */
 
+#undef LOG_COMPILE_TIME_LEVEL
+#define LOG_COMPILE_TIME_LEVEL LOG_LEVEL_WARN
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply system/src/system_task.cpp]: Redefined.

File: hal/network/util/simple_ntp_client_detail.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (sergeuz) [reply hal/network/util/simple_ntp_client_detail.h:93]: The description you have added is just awesome :pray:

File: hal/network/util/simple_ntp_client.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sergeuz) [reply hal/network/util/simple_ntp_client.cpp:154]: `isOwnSock_` is even better :+1:

File: hal/network/util/simple_ntp_client.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (sergeuz) [reply hal/network/util/simple_ntp_client.cpp]: True :smiley:

File: hal/network/util/simple_ntp_client.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/network/util/simple_ntp_client.cpp:154]: Changed to that :)

**Pull request merged**